### PR TITLE
Add support for more power levels for HVX

### PR DESCRIPTION
**Pull request by suyogsarda**
This commit adds support for 6 power levels for HVX.

### CODE REVIEW

File: src/runtime/hexagon_remote/halide_remote.cpp [hunk 1]
```diff
@@ -157,19 +168,37 @@ int halide_hexagon_remote_set_performance(
...
-    request.mips_bw.set_bus_bw      = set_bus_bw;
-    request.mips_bw.bwBytePerSec    = ((uint64_t) bwMegabytesPerSec) << 20;
-    request.mips_bw.busbwUsagePercentage = busbwUsagePercentage;
-    request.mips_bw.set_latency     = set_latency;
-    request.mips_bw.latency         = latency;
-    retval = HAP_power_set(NULL, &request);
-    if (0 != retval) {
-        log_printf("HAP_power_set(HAP_power_set_mips_bw) failed (%d)\n", retval);
-        return -1;
+    if (use_dlopenbuf()) {
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/halide_remote.cpp]: At best, this seems coincidentally correct (using dlopenbuf as a proxy for the v2 of this being present). Is there some other way to determine if HAP_power_set_DCVS_v2 is available? Maybe it should simply try HAP_power_set with HAP_power_set_DCVS_v2, and if it fails, then use the old one?

Also, are these actually mutually exclusive? It seems like they're setting different things (except latency).

File: src/runtime/hexagon_remote/halide_remote.cpp [hunk 2]
```diff
@@ -157,19 +168,37 @@ int halide_hexagon_remote_set_performance(
...
+        request.dcvs_v2.dcvs_option = HAP_power_dcvs_v2_payload::HAP_DCVS_V2_POWER_SAVER_MODE;
+        request.dcvs_v2.set_dcvs_params = TRUE;
+        request.dcvs_v2.dcvs_params.min_corner = HAP_DCVS_VCORNER_DISABLE;
+        request.dcvs_v2.dcvs_params.max_corner = HAP_DCVS_VCORNER_DISABLE;
+        request.dcvs_v2.dcvs_params.target_corner = voltageCorner[powerlevel];
+        request.dcvs_v2.set_latency     = set_latency;
+        request.dcvs_v2.latency         = latency;
+        retval = HAP_power_set(NULL, &request);
+        if (0 != retval) {
+            log_printf("HAP_power_set(HAP_power_set_mips_bw) failed (%d)\n", retval);
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/halide_remote.cpp]: This should probably change HAP_power_set_mips_bw -> HAP_power_dcvs_v2

File: src/runtime/hexagon_remote/halide_remote.cpp [hunk 3]
```diff
@@ -145,10 +145,21 @@ int halide_hexagon_remote_set_performance(
...
     unsigned int busbwUsagePercentage,
     int set_latency,
-    int latency) {
+    int latency,
+    int powerlevel) {
 
     HAP_power_request_t request;
-
+    const uint32_t numPowerLevels = 7;
+    const HAP_dcvs_voltage_corner_t voltageCorner[numPowerLevels]
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/halide_remote.cpp]: Seems safer/easier to remote numPowerLevels above, change this to:

    const HAP_dcvs_voltage_corner_t voltage_corners[] = {
        ...
        num_power_levels,
    };

File: src/runtime/hexagon_remote/halide_remote.cpp [hunk 4]
```diff
@@ -157,19 +168,37 @@ int halide_hexagon_remote_set_performance(
...
-    request.mips_bw.latency         = latency;
-    retval = HAP_power_set(NULL, &request);
-    if (0 != retval) {
-        log_printf("HAP_power_set(HAP_power_set_mips_bw) failed (%d)\n", retval);
-        return -1;
+    if (use_dlopenbuf()) {
+        memset(&request, 0, sizeof(HAP_power_request_t));
+        request.type = HAP_power_set_DCVS_v2;
+        request.dcvs_v2.dcvs_enable = TRUE;
+        request.dcvs_v2.dcvs_option = HAP_power_dcvs_v2_payload::HAP_DCVS_V2_POWER_SAVER_MODE;
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/halide_remote.cpp]: Do we need to be able to control this? It seems like it might have a big impact on performance?

Comment (suyogsarda) [reply src/runtime/hexagon_remote/halide_remote.cpp]: We haven't observed any performance regression for hexagon_benchmark with HAP_DCVS_V2_POWER_SAVER_MODE, although there are other modes too (including PERF mode, which might penalize in terms of power). How do we combine the two - dcvs_option and target_corner? Or should we just leave it without specifying dcvs_option? Note that HAP_DCVS_V2_POWER_SAVER_MODE is the default option for all the clients participating in dcvs, and hence specified so.

File: src/runtime/hexagon_remote/halide_remote.cpp (hunk 1 above)
Comment (suyogsarda) [reply src/runtime/hexagon_remote/halide_remote.cpp]: I tested the "try old if new fails" but it doesn't set DCVS correctly. using dlopenbuf() to differentiate the two seems the only option for now, although it doesn't seem not clean logically.

**Comment by suyogsarda** [comment]
I think the mapping of enum to new values doesn't break the old code as dlopenbuf() is being used to differentiate between the two in runtime. Checked backward compatibility too on the device not supporting DCVS, runs perfectly fine. However, i can make this a new api as set_performance_v2() if it seems so. However, user may have to change the api to v2 in his code if he wants to use DCVS. Current implementation doesn't need user to bother about it, while providing more options for power. 

I can change to separate api v2 if it seems unavoidable.

File: src/runtime/hexagon_remote/halide_remote.cpp (hunk 1 above)
Comment (dsharletg) [reply src/runtime/hexagon_remote/halide_remote.cpp]: There *must* to be a way to do this in the HAP API. If there isn't, Qualcomm should add one. At the very least, trying the new one and falling back to the old one really should work. Why doesn't that work?

**Comment by dsharletg** [comment]
The layer we need backwards compatibility at is the Halide remote runtime library.

What should happen is that in addition to renaming the API, we should also add a shim for the old API to the new API. There are a few examples here: https://github.com/halide/Halide/blob/master/src/runtime/hexagon_remote/host_shim.cpp#L14

This way, old code will run with the new runtime (by calling the shim).

@ronlieb has also done this in the past and might be able to help?

**Comment by ronlieb** [comment]
Hi Dillon,

I will get with Suyog in the morning to help with that ,

 

Thanks,

Ron

 

---

Qualcomm Innovation Center, Inc. is a member of Code Aurora Forum, hosted by The Linux Foundation

**Comment by steven-johnson** [comment]
This has been sitting open (without any action) for a good while now. Is it still an ongoing concern?

**Comment by suyogsarda** [comment]
Will reopen after some rework.

**Pull request closed without merging**